File: structures/stack/stack.py
class Stack:
    class Node:
        def __init__(self, element):
            self.element = element
            self.next = None

    def __init__(self):
        self.top = None
        self.capacity = 0

    def push(self, value):
        new_node = self.Node(value)
        new_node.next = self.top
        self.top = new_node
        self.capacity += 1

    def pop(self):
        if self.is_empty():
            raise IndexError("A pilha está vazia!")
        self.top = self.top.next
        self.capacity -= 1

    def peek(self):
        if self.is_empty():
            raise IndexError("A pilha está vazia!")
        return self.top.element

    def size(self):
        return self.capacity

    def is_empty(self):
        return self.top is None

    def clear(self):
        self.top = None
        self.capacity = 0

File: structures/stack/Stack.java
package stack;
import java.util.NoSuchElementException;

public class Stack {

    private class Node {
        public Integer element;
        public Node next;

        public Node(Integer element) {
            this.element = element;
            next = null;
        }
    }

    private Node top;
    private int capacity;

    public Stack() {
        top = null;
        capacity = 0;
    }

    public void push(int value) {
        Node n = new Node(value);
        n.next = top;
        top = n;
        capacity++;
    }

    public void pop() {
        if (isEmpty()) {
            throw new NoSuchElementException("Stack is empty!");
        }
        top = top.next;
        capacity--;
    }

    public int peek() {
        if (isEmpty()) {
            throw new NoSuchElementException("Stack is empty!");
        }
        return top.element;
    }

    public int size() {
        return capacity;
    }

    public boolean isEmpty() {
        return top == null;
    }

    public void clear() {
        top = null;
        capacity = 0;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        
        if (capacity == 0) {
            sb.append("Stack: No Element");
        } else {
            Node current = top;
            sb.append("Stack: ");
            while (current != null) {
                sb.append(current.element).append(" ");
                current = current.next;
            }
        }
        
        return sb.toString();
    }
    
}
